pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -1309,3 +1332,90 @@ func initializeAnnotations(cd *hivev1.ClusterDeployment) {
 		cd.Annotations = map[string]string{}
 	}
 }
+
+// managePullSecrets() takes pull secretes in to account and returns the merged pull request
+func (r *ReconcileClusterDeployment) managePullSecrets(cd *hivev1.ClusterDeployment, cdLog log.FieldLogger) (string, error) {
+
+	// For code readability let's call the pull secret in cluster deployment config as local pull secret
+	localPullSecret, _ := controllerutils.LoadSecretData(r.Client, cd.Spec.PullSecret.Name, cd.Namespace, corev1.DockerConfigJsonKey)
+
+	// Check if global pull secret from env as it comes from hive config
+	globalPullSecret := os.Getenv("GLOBAL_PULL_SECRET")
+
+	switch {
+	case globalPullSecret != "" && localPullSecret != "":
+		// Merge local pullSecret and globalPullSecret. If both pull secrets have same registry name
+		// then the merged pull secret will have registry secret from local pull secret
+		pullSecret, err := controllerutils.MergeJsons(globalPullSecret, localPullSecret, cdLog)
+		if err != nil {
+			errMsg := "unable to merge global pull secret with local pull secret"
+			cdLog.WithError(err).Error(errMsg)
+			return "", fmt.Errorf("%s : %v", errMsg, err)
+		}
+		return pullSecret, nil
+	case globalPullSecret != "":
+		return globalPullSecret, nil
+	case localPullSecret != "":
+		return localPullSecret, nil
+	default:
+		errMsg := "At least one of pull request need to be available"
+		cdLog.Error(errMsg)
+		return "", fmt.Errorf("%s", errMsg)
+	}
+}
+
+// updatePullSecretObj() adds pull secret information in cluster deployment and cluster deployment status
+func (r *ReconcileClusterDeployment) updatePullSecretInfo(pullSecret string, cd *hivev1.ClusterDeployment, cdLog log.FieldLogger) error {
```

Comment:
I find this function a little overly complex. I suggest the following code abstract.

```
existingPullSecret = findExisting()
if existingPullSecret != nil {
  if hash(pullSecret) == hash(existingPullSecret.Spec.Data["..."]) {
    return nil
  }
}
createNewSecret()
defer deleteNewSecretIfFailure()
deleteExistingSecretIfExists()
updateClusterDeploymentPullSecret()
```

Reply:
Sure I will simplify it. However I don't think we need `deleteExistingSecretIfExists()` as we will be using same `pullSecretName` when we  `Update(context.TODO(), pullSecretObj)`

Reply:
So the question is should we delete existing secrete and then always create new one or update the existing one if possible. My guess is updating might be efficient but not sure. CC @dgoodwin

Reply:
I would expect it to re-use the secret once one was created and successfully set on the clusterdeployment. What about:

existingPullSecret = findExisting()
if existingPullSecret != nil {
  if hash(pullSecret) == hash(existingPullSecret.Spec.Data["..."]) {
    return nil
  }
  return updatePullSecret()
}
createNewSecret()
updateClusterDeploymentPullSecret()
defer deleteNewSecretIfFailure() // if the update on clusterdeploymentfailed

Reply:
I would expect it to never re-use the secret, so that the controller never has to reason about whether the secret stored in the local cache is an out-of-date version. The clusterdeployment is always updated after the secret is created. If there is a failure before the secret is created, then the clusterdeployment is not updated.

Reply:
Works for me.

Reply:
The other wrinkle with re-using the secret is that any new pod created may use the previous version of the secret rather than the new, correct version. Kube only offers eventual consistency, so there is no guarantee that the pod created will see the most recent version--which will be exacerbated by creating the installer job very quickly after updating the secret.

Reply:
This should be fixed now.